crypto/src/fixpoint.rs

```diff
@@ -181,3 +172,645 @@ impl U128x128 {
         U128x128(self.0.saturating_sub(rhs.0))
     }
 }
+
+pub struct U128x128Var {
+    pub limbs: [UInt64<Fq>; 4],
+}
+
+impl AllocVar<U128x128, Fq> for U128x128Var {
+    fn new_variable<T: std::borrow::Borrow<U128x128>>(
+        cs: impl Into<ark_relations::r1cs::Namespace<Fq>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: ark_r1cs_std::prelude::AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+        let inner: U128x128 = *f()?.borrow();
+
+        let bytes = inner.to_bytes();
+        // The U128x128 type uses a big-endian encoding
+        let limb_3 = u64::from_be_bytes(bytes[0..8].try_into().unwrap());
+        let limb_2 = u64::from_be_bytes(bytes[8..16].try_into().unwrap());
+        let limb_1 = u64::from_be_bytes(bytes[16..24].try_into().unwrap());
+        let limb_0 = u64::from_be_bytes(bytes[24..32].try_into().unwrap());
+
+        let limb_0_var = UInt64::new_variable(cs.clone(), || Ok(limb_0), mode)?;
+        let limb_1_var = UInt64::new_variable(cs.clone(), || Ok(limb_1), mode)?;
+        let limb_2_var = UInt64::new_variable(cs.clone(), || Ok(limb_2), mode)?;
+        let limb_3_var = UInt64::new_variable(cs, || Ok(limb_3), mode)?;
+
+        Ok(Self {
+            limbs: [limb_0_var, limb_1_var, limb_2_var, limb_3_var],
+        })
+    }
+}
+
+impl R1CSVar<Fq> for U128x128Var {
+    type Value = U128x128;
+
+    fn cs(&self) -> ark_relations::r1cs::ConstraintSystemRef<Fq> {
+        self.limbs[0].cs()
+    }
+
+    fn value(&self) -> Result<Self::Value, ark_relations::r1cs::SynthesisError> {
+        let x0 = self.limbs[0].value()?;
+        let x1 = self.limbs[1].value()?;
+        let x2 = self.limbs[2].value()?;
+        let x3 = self.limbs[3].value()?;
+
+        let mut bytes = [0u8; 32];
+        bytes[0..8].copy_from_slice(x3.to_be_bytes().as_ref());
+        bytes[8..16].copy_from_slice(x2.to_be_bytes().as_ref());
+        bytes[16..24].copy_from_slice(x1.to_be_bytes().as_ref());
+        bytes[24..32].copy_from_slice(x0.to_be_bytes().as_ref());
+
+        Ok(Self::Value::from_bytes(bytes))
+    }
+}
+
+impl U128x128Var {
+    pub fn checked_add(
+        self,
+        rhs: &Self,
+        cs: ConstraintSystemRef<Fq>,
+    ) -> Result<U128x128Var, SynthesisError> {
+        // x = [x0, x1, x2, x3]
+        // x = x0 + x1 * 2^64 + x2 * 2^128 + x3 * 2^192
+        // y = [y0, y1, y2, y3]
+        // y = y0 + y1 * 2^64 + y2 * 2^128 + y3 * 2^192
+        let x0 = convert_uint64_to_fqvar(&self.limbs[0]);
```

Comment:
For later it could make sense to lazily/optionally store the limbs as field elements on the `U128x128Var` struct in addition to `UInt64`s